// @nrwl/angular/src/builders/module-federation-dev-server/module-federation-dev-server.impl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeModuleFederationDevServerBuilder = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const ngcli_adapter_1 = require("nx/src/adapter/ngcli-adapter");
const architect_1 = require("@angular-devkit/architect");
const path_1 = require("path");
const webpack_server_impl_1 = require("../webpack-server/webpack-server.impl");
const fs_1 = require("fs");
function getDynamicRemotesIfExists(context, p) {
    // check for dynamic remotes
    // we should only check for dynamic based on what we generate
    // and fallback to empty array
    const standardPathToGeneratedMFManifestJson = (0, path_1.join)(context.workspaceRoot, p.sourceRoot, 'assets/module-federation.manifest.json');
    if (!(0, fs_1.existsSync)(standardPathToGeneratedMFManifestJson)) {
        return [];
    }
    const moduleFederationManifestJson = (0, fs_1.readFileSync)(standardPathToGeneratedMFManifestJson, 'utf-8');
    if (moduleFederationManifestJson) {
        // This should have shape of
        // {
        //   "remoteName": "remoteLocation",
        // }
        const parsedManifest = JSON.parse(moduleFederationManifestJson);
        if (!Object.keys(parsedManifest).every((key) => typeof key === 'string' && typeof parsedManifest[key] === 'string')) {
            return [];
        }
        return Object.entries(parsedManifest).reduce((remotesArray, [remoteName, remoteLocation]) => [
            ...remotesArray,
            [remoteName, remoteLocation],
        ], []);
    }
    return [];
}
function executeModuleFederationDevServerBuilder(schema, context) {
    const workspaceConfig = (0, devkit_1.readAllWorkspaceConfiguration)();
    const p = workspaceConfig.projects[context.target.project];
    const mfConfigPath = (0, path_1.join)(context.workspaceRoot, p.root, 'module-federation.config.js');
    let mfeConfig;
    try {
        mfeConfig = require(mfConfigPath);
    }
    catch (_a) {
        throw new Error(`Could not load ${mfConfigPath}. Was this project generated with "@nrwl/angular:host"?`);
    }
    const options = tslib_1.__rest(schema, []);
    let unparsedRemotes = mfeConfig.remotes.length > 0 ? mfeConfig.remotes : [];
    const dynamicRemotes = getDynamicRemotesIfExists(context, p);
    unparsedRemotes = [...unparsedRemotes, ...dynamicRemotes];
    const remotes = unparsedRemotes.map((a) => (Array.isArray(a) ? a[0] : a));
    const devServeRemotes = !options.devRemotes
        ? []
        : Array.isArray(options.devRemotes)
            ? options.devRemotes
            : [options.devRemotes];
    for (const remote of remotes) {
        const isDev = devServeRemotes.includes(remote);
        const target = isDev ? 'serve' : 'serve-static';
        (0, ngcli_adapter_1.scheduleTarget)(context.workspaceRoot, {
            project: remote,
            target,
            configuration: context.target.configuration,
            runOptions: {},
            executor: context.builder.builderName,
        }, options.verbose).then((obs) => {
            obs.toPromise().catch((err) => {
                throw new Error(`Remote '${remote}' failed to serve correctly due to the following: \r\n${err.toString()}`);
            });
        });
    }
    return (0, webpack_server_impl_1.executeWebpackServerBuilder)(options, context);
}
exports.executeModuleFederationDevServerBuilder = executeModuleFederationDevServerBuilder;
exports.default = (0, architect_1.createBuilder)(executeModuleFederationDevServerBuilder);
